### ISSUE TITLE
WPOLYMER not working after start of time stepping

### ISSUE THREAD

**Issue by EdmundStephens** [OPM/opm-simulators#5684]
[v2024.04] The attached example pair shows a polymer model run. Using WPOLYMER to assign polymer injection to a well after start of time stepping does not work, in fact it seems to disable the polymer mode altogether. The workaround is to use separate injectors for water and polymer water, this is possible in simple test cases but not handy to arrange in complex cases. 

* R1_SS_poly1f ... uses 2 injectors I1 (water) and I2 (water with polymer) where these are switched after 2 years
* R1_SS_poly1f1 ... uses 1 injector and WPOLYMER assignment at 2 years ... model does not work

I had the chance to run the R1_SS_poly1f1 in tNavigator which resulting in the correct polymer injection.

**Comment by GitPaean** [comment]
Hi, 

Thanks for reaching to us. 

There are a few keywords that we do not support, 
![image](https://github.com/user-attachments/assets/e0577c08-d11c-4c0d-b1c5-1b91753204fb)

After commenting out the few lines, it looks running fine with both the 1f and 1f1 cases. 

The results of the 1f1 case can be seen from the following file. 
[summary.pdf](https://github.com/user-attachments/files/17461640/summary.pdf)

Can you give more details about your finding, regarding that polymer mode is disabled?

And also, which version of OPM-flow are using?

Thanks. 

Best, 
Kai

**Comment by EdmundStephens** [comment]
I'm using version 2024.04 and still see this bug (after removing OLDTRAN, RSCONST and GCONTOL). 

Runs attached with all includes, grid and summary output data. I simplified the models to help home in on this issue with three cases: 
* Ref1 is a reference case with simple water flood
* Poly1 uses 1 injector well and changes the polymer concentration via WPOLYMER at t=2 years
* Poly2 uses 2 injectors with I1=water and I2=water+polymer and switches I2 on at t=2 years

[bug_polymer.zip](https://github.com/user-attachments/files/17467203/bug_polymer.zip)

See images below. There are two issues: 
1. In case Poly1, WPOLYMER is not working. The case should show the polymer flood but does not; in fact it gives the same result as well as run performance as the Ref1 which has no polymer ... in the output all MB(P) and CNV(P) are exactly zero. Poly2 uses two wells with pre-set polymer, then switches these via WELOPEN ... worth noting that this is a successful workaround, albeit could be hard to arrange in more general models, but it would be better to identify and resolve the problem. 
2. In case Poly2, the result is correct but run times are on long side, the attached zip includes the settings used via --parameter=file=default.param. In the second plot below the polymer run has run time ~ 880 seconds compared to the water flood ~9 seconds. This differential (x 100) is much larger than it should be: the solver is repeatedly failing and reducing time step size, especially after 2026 when there is polymer flowing into the producer well (presumable sensitivity between well performance and fluid inflow mobilities). I have tried at length to find parameter choices that make a difference but without success ... is there any particular solver parameters or other settings that help ?

![image](https://github.com/user-attachments/assets/d52e339d-bc7d-4f77-bf93-6d6a3041081d)
![image](https://github.com/user-attachments/assets/e4576a8b-3b4b-4a70-a311-23e25eb16800)

**Comment by GitPaean** [comment]
Thanks. Please look at my running with the case 1f1, https://github.com/user-attachments/files/17461640/summary.pdf

It looks like `1f1` shows desirable results,  (WOPR:P1 from the summary.pdf) (`1f1` is the case you claim not working. )

![image](https://github.com/user-attachments/assets/60592803-b5a8-48b1-bd58-274dd15448ab)


I recompiled a 2024.4 release version, 

it looks like POLY1F case that you provided is also working, the following is a unfinished running (green line is my running)

![image](https://github.com/user-attachments/assets/1fa42630-2fd2-4003-8a51-8a1624905258)

I do not work with installed release version, so I am not sure what happened with your running. 

What is the easiest way to test release build? @akva2 


Regarding the long running time, it is mostly due to convergence issue, which needs more close investigation and I do not have capacity for it, sorry about that.

**Comment by EdmundStephens** [comment]
Kai, Thank you so much, this is a curious one it might even be some incompatibility in the release versions of some of the common libraries, so that you don't see it in your development version. Weird. I am running on Ubuntu on Windows 10 via WSL 2. I get the same issue on my laptop (MS Surface Pro 7, Intel i7) and workstation (custom build with Ryzen threadripper 32 core), fully updated on both Windows and Ubuntu sides, although still showing some very minor differences between the platforms. 
> Laptop: Ubuntu 22.04.5 LTS (GNU/Linux 5.10.16.3-microsoft-standard-WSL2 x86_64)
> Workstation: Ubuntu 22.04.5 LTS (GNU/Linux 5.10.102.1-microsoft-standard-WSL2 x86_64)

**Comment by GitPaean** [comment]
> although still showing some very minor differences between the platforms.

So on both platforms, the polymer mode does not work for POLY1F case for you. Then we need to check the release build. 

I build a release for 2024.04 locally, the polymer mode works for POLY1F.

**Comment by akva2** [comment]
yeah, this is a new problem. i'll have to fix up the packaging, for some reason a shared object has ended up in both of the packages. for now, unless you need it, remove the octave-resinsight package.

**Comment by EdmundStephens** [comment]
Just FYI, I did sudo apt autoremove and full-upgrade on both machines, then removed octave-resinsight. Unfortunately the problem persists: Poly1f fails, Poly2f runs but slowly. 

Grateful for your attention, with any luck this might expose and help you resolve some sort of nasty internal issue. On the solver side I have noticed for example that the solver is slow whenever there is a 4th component including thermal runs.

**Comment by GitPaean** [comment]
> > although still showing some very minor differences between the platforms.
> 
> So on both platforms, the polymer mode does not work for POLY1F case for you. Then we need to check the release build.
> 
> I build a release for 2024.04 locally, the polymer mode works for POLY1F.

One small update. I tested in build release with the following command, 

```
sudo apt-add-repository ppa:opm/ppa
sudo apt-get update
sudo apt-get install libopm-simulators-bin

kaib@kai-ws:~$ which flow
/usr/bin/flow
kaib@kai-ws:~$ flow --version
flow 2024.04
```

with this build, the polymer mode works with POLY1F.  

![image](https://github.com/user-attachments/assets/271224bc-2bf4-4e27-be31-f078f04179e3)

It is out of knowledge what happened with your installation.

**Comment by EdmundStephens** [comment]
Well I tried these and no luck so I'm not sure what that leave you if you cannot reproduce the issue. Seems very odd to me. Can you push this issue around your developer base see if anyone has ideas. Thing is the student this comes from sees the same thing and has a completely fresh installation of Ubuntu made last week for this project, so I dont think a complete reinstall will necessarily be an answer. 

```
ed@DESKTOP-8VC26F1:~$ sudo apt-add-repository ppa:opm/ppa
Repository: 'deb https://ppa.launchpadcontent.net/opm/ppa/ubuntu/ jammy main'
Description:
The Open Porous Media (OPM) initiative provides a set of open-source tools centered around the simulation of flow and transport of fluids in porous media. The goal of the initiative is to establish a sustainable environment for the development of an efficient and well-maintained software suite.
More info: https://launchpad.net/~opm/+archive/ubuntu/ppa
Adding repository.
Press [ENTER] to continue or Ctrl-c to cancel.
Found existing deb entry in /etc/apt/sources.list.d/opm-ubuntu-ppa-jammy.list
Adding deb entry to /etc/apt/sources.list.d/opm-ubuntu-ppa-jammy.list
Found existing deb-src entry in /etc/apt/sources.list.d/opm-ubuntu-ppa-jammy.list
Adding disabled deb-src entry to /etc/apt/sources.list.d/opm-ubuntu-ppa-jammy.list
Adding key to /etc/apt/trusted.gpg.d/opm-ubuntu-ppa.gpg with fingerprint 91121C48A10BAC661586B42A50BDC27473618B59
Hit:1 https://ppa.launchpadcontent.net/opm/ppa/ubuntu jammy InRelease
Hit:2 http://archive.ubuntu.com/ubuntu jammy InRelease
Hit:3 http://security.ubuntu.com/ubuntu jammy-security InRelease
Hit:4 http://archive.ubuntu.com/ubuntu jammy-updates InRelease
Hit:5 http://archive.ubuntu.com/ubuntu jammy-backports InRelease
Reading package lists... Done
ed@DESKTOP-8VC26F1:~$ sudo apt-get update
Hit:1 https://ppa.launchpadcontent.net/opm/ppa/ubuntu jammy InRelease
Hit:2 http://security.ubuntu.com/ubuntu jammy-security InRelease
Hit:3 http://archive.ubuntu.com/ubuntu jammy InRelease
Hit:4 http://archive.ubuntu.com/ubuntu jammy-updates InRelease
Hit:5 http://archive.ubuntu.com/ubuntu jammy-backports InRelease
Reading package lists... Done
ed@DESKTOP-8VC26F1:~$ sudo apt-get install libopm-simulators-bin
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
libopm-simulators-bin is already the newest version (2024.04-1~jammy).
0 upgraded, 0 newly installed, 0 to remove and 1 not upgraded.
ed@DESKTOP-8VC26F1:~$ which flow
/usr/bin/flow
ed@DESKTOP-8VC26F1:~$ flow --version
flow 2024.04
ed@DESKTOP-8VC26F1:~$
```

**Comment by GitPaean** [comment]
The only difference that I can see is that you are using  Ubuntu 22.04 and I am using Ubuntu 24.04.

**Comment by GitPaean** [comment]
another update, I use virtual machine of Ubuntu 22.04 and installed a OPM-flow 24.04, still I could not reproduce your issue.  

![image](https://github.com/user-attachments/assets/b20d6178-5cab-4bab-be42-6a3b95b35924)

there is something really hard to answer from my end.

**Comment by EdmundStephens** [comment]
I get a message in the login phase

> Windows Subsystem for Linux is now available in the Microsoft Store!
> You can upgrade by running 'wsl.exe --update' or by visiting https://aka.ms/wslstorepage
> Installing WSL from the Microsoft Store will give you the latest WSL updates, faster.
> For more information please visit https://aka.ms/wslstoreinfo

which ties to this discussion, so I am mid process of updating and will get back to you shortly

First on Windows command prompt (Admin)
> wsl.exe --update

Then in the Ubuntu shell
$ sudo apt remove snapd
$ sudo do-release-upgrade -d

... ongoing (its a rather long process)

**Comment by EdmundStephens** [comment]
Long install process followed by restarting WSL and Ubuntu
> wsl.exe --shutdown

> $ lsb_release -a
> No LSB modules are available.
> Distributor ID: Ubuntu
> Description:    Ubuntu 24.04.1 LTS
> Release:        24.04
> Codename:       noble

and reinstall OPM
> sudo apt-add-repository ppa:opm/ppa
> sudo apt-get update
> apt-cache search opm-simulators
> sudo apt-get install mpi-default-bin
> sudo apt-get install libopm-simulators-bin

... and ... drumroll ... no change in the result for poly1f run. 

Sorry ...

**Comment by GitPaean** [comment]
I reproduced your problem now. 

For your case, when I ran with 4 or more processes, poly1f does not show correct behavoir. 

When I ran with 1 or 2 processes, poly1f shows desired behavoir. 

Both the development branch and release branch have the same problem. 

So there is a bug for the parallel running related to your polymer case. 

For now I can only suggest you run the cases with 1 or 2 processes.   I am not sure whether we can prioritize the fix because there is not much polymer development at the moment.  But we will keep it in mind. 

Please let us know whether you see desired results with 1 or 2 processes?

**Comment by GitPaean** [comment]
Checking the output, it also confirms that when the polymer mode not working, there is indeed no polymer existing in the field.

**Comment by GitPaean** [comment]
It is related to the parallel distribution of the wells after parsing the DATA file. We will see whether we can fix it easily.  You will hear from us if there is any update.

**Comment by EdmundStephens** [comment]
Great - that is some impressive detective work there. For the time being the workaround (that you can circulate to interested users) is to have duplicate wells as the polymer injector and switch these via WELOPEN. 

Would you like me to submit the more general polymer solver issue as a separate issue ? To me it seems related to other circumstances where the 4th component is used, including a lot of my Geothermal work (black oil TEMP model) and CCS in black oil with the Solvent method (per Matthias & Oldenberg).

**Comment by GitPaean** [comment]
> Would you like me to submit the more general polymer solver issue as a separate issue ? To me it seems related to other circumstances where the 4th component is used, including a lot of my Geothermal work (black oil TEMP model) and CCS in black oil with the Solvent method (per Matthias & Oldenberg).

sure, that will be great!

I am in progress in debugging the current case at the moment.

**Comment by EdmundStephens** [comment]
Ok will do. I think I will check the solver on the Thermal and CCS cases, if these have a common pattern I can whizz up a set of simplified test decks to you guys to target in debugging.

**Comment by GitPaean** [comment]
>  Ok will do. I think I will check the solver on the Thermal and CCS cases, if these have a common pattern I can whizz up a set of simplified test decks to you guys to target in debugging.

much appreciated!

**Comment by GitPaean** [comment]
@EdmundStephens  thanks for reporting this issue. With the help of some colleague, we figured out the issue and proposed a fix  https://github.com/OPM/opm-common/pull/4277 . 

We are in the final stage of publishing new release of 2024.10. Not sure whether the fix can enter this release. It if did not, the fix can only be available for the simulator compiled from the source code. 

Will keep you updated with the issue.

**Comment by GitPaean** [comment]
My colleague just confirmed that the fix will go in the 2024.10 release.  So within about a week, you can have a new version of OPM-flow will not have the same issue (polymer not working in parallel running. )
